From 3e4ac4c60929ea3c9d4d1f3f436b6b3c2d8428ad Mon Sep 17 00:00:00 2001
From: Ricardo Neri <ricardo.neri@ti.com>
Date: Mon, 7 Mar 2011 12:47:09 -0600
Subject: ASoC: OMAP4: Configure sample rate and format in HDMI DAI

Add support to configure sample rate and format of the HDMI DAI.
It includes support for 32, 44.1 and 48kHz sample rates and 16-bit
and 24-bit (in 32-bit word) sample formats.

Change-Id: I2acae4d92723ff2dfb1a9c816210f1511a2da8bb
Signed-off-by: Ricardo Neri <ricardo.neri@ti.com>
---
 sound/soc/omap/omap-hdmi.c |   26 +++++++++++++++++++-------
 sound/soc/omap/sdp4430.c   |    5 +++--
 2 files changed, 22 insertions(+), 9 deletions(-)

Index: kernel/sound/soc/omap/omap-hdmi.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-hdmi.c	2011-04-07 12:11:22.000000000 +0900
+++ kernel/sound/soc/omap/omap-hdmi.c	2011-04-07 12:22:12.000000000 +0900
@@ -43,10 +43,14 @@
 #endif
 
 #define CONFIG_HDMI_NO_IP_MODULE
-#define OMAP_HDMI_RATES	(SNDRV_PCM_RATE_48000)
+#define OMAP_HDMI_RATES        (SNDRV_PCM_RATE_32000 | \
+	SNDRV_PCM_RATE_44100 | SNDRV_PCM_RATE_48000)
+
+
+/* Support for 16 and 24 bits */
+#define OMAP_HDMI_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | \
+	SNDRV_PCM_FMTBIT_S24_LE)
 
-/* Currently, we support only 16b samples at HDMI */
-#define OMAP_HDMI_FORMATS (SNDRV_PCM_FMTBIT_S16_LE)
 
 
 #include <plat/hdmi_lib.h>
@@ -197,13 +201,23 @@
 
 	switch (params_format(params)) {
 	case SNDRV_PCM_FORMAT_S16_LE:
-	case SNDRV_PCM_FORMAT_S32_LE:
-		omap_hdmi_dai_dma_params.data_type = OMAP_DMA_DATA_TYPE_S32;
+		err = hdmi_configure_audio_sample_size(HDMI_SAMPLE_16BITS);
+		break;
+	case SNDRV_PCM_FORMAT_S24_LE:
+		err = hdmi_configure_audio_sample_size(HDMI_SAMPLE_24BITS);
 		break;
 
 	default:
-		err = -EINVAL;
+		return -EINVAL;
 	}
+	if (err)
+		return err;
+
+	err = hdmi_configure_audio_sample_freq(params_rate(params));
+	if (err)
+		return err;
+
+	omap_hdmi_dai_dma_params.data_type = OMAP_DMA_DATA_TYPE_S32;
 	omap_hdmi_dai_dma_params.packet_size = 0x20;
 
 	snd_soc_dai_set_dma_data(dai, substream,
Index: kernel/sound/soc/omap/sdp4430.c
===================================================================
--- kernel.orig/sound/soc/omap/sdp4430.c	2011-04-07 10:34:49.000000000 +0900
+++ kernel/sound/soc/omap/sdp4430.c	2011-04-07 12:16:14.000000000 +0900
@@ -554,8 +554,9 @@
 		.stream_name = "Playback",
 		.channels_min = 2,
 		.channels_max = 8,
-		.rates = SNDRV_PCM_RATE_48000,
-		.formats = SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S32_LE,
+		.rates = SNDRV_PCM_RATE_32000 | SNDRV_PCM_RATE_44100 |
+			SNDRV_PCM_RATE_48000,
+		.formats = SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE,
 	},
 },
 };
